File: 5-saga-orchestration-and-choreography/1-java/inventory-service/src/main/java/com/starci/inventory/StockService.java
package com.starci.inventory;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

/**
 * StockService — core saga choreography logic for inventory reservation.
 *
 * <p>Logic:
 * <ol>
 *   <li>{@link #tryFulfill} — idempotency check, stock validation, decrement, emit reply event.</li>
 *   <li>{@link #handleSagaEvent} — route ORDER_CREATED / PAYMENT_CAPTURED events to tryFulfill.</li>
 * </ol>
 */
@Service
public class StockService {

    private static final Logger log = LoggerFactory.getLogger(StockService.class);

    private final ProductRepository products;
    private final FulfillmentRepository fulfillments;
    private final KafkaTemplate<String, Object> kafka;

    /**
     * Constructor injection.
     *
     * @param products     MongoDB repository for product stock documents
     * @param fulfillments MongoDB repository for idempotency fulfillment records
     * @param kafka        Spring Kafka producer template for saga reply events
     */
    public StockService(ProductRepository products,
                        FulfillmentRepository fulfillments,
                        KafkaTemplate<String, Object> kafka) {
        this.products = products;
        this.fulfillments = fulfillments;
        this.kafka = kafka;
    }

    /**
     * Attempt to fulfill the stock reservation for an order.
     *
     * <p>Logic:
     * <ul>
     *   <li>Idempotency: skip if fulfillment record already exists for this orderId.</li>
     *   <li>Validate: emit INVENTORY_OUT_OF_STOCK if product missing or stock insufficient.</li>
     *   <li>Commit: decrement stock in MongoDB, save fulfillment record, emit INVENTORY_OK.</li>
     * </ul>
     *
     * @param orderId   order identifier from the saga event
     * @param productId product to reserve stock for
     * @param quantity  units requested
     * @return result object describing the outcome
     */
    public InventoryCheckResult tryFulfill(Long orderId, Long productId, Integer quantity) {
        // Idempotency check: skip duplicate events that have already been processed.
        if (fulfillments.existsById(orderId)) {
            InventoryCheckResult result = new InventoryCheckResult();
            result.setOk(true);
            result.setOrderId(orderId);
            result.setProductId(productId);
            result.setQuantity(quantity);
            result.setStatus("ALREADY_FULFILLED");
            result.setMessage("Order has already been fulfilled");
            return result;
        }

        // Load product from MongoDB — Integer id matches the seeded values (1, 2).
        Optional<Product> productOpt = products.findById(productId.intValue());

        if (productOpt.isEmpty() || productOpt.get().getStock() < quantity) {
            // Emit compensating event so the order-service can cancel the order.
            Map<String, Object> outOfStockEvent = new HashMap<>();
            outOfStockEvent.put("event", "INVENTORY_OUT_OF_STOCK");
            outOfStockEvent.put("orderId", orderId);
            outOfStockEvent.put("productId", productId);
            kafka.send(SagaConstants.TOPIC, outOfStockEvent);
            log.info("Out of stock for order {}", orderId);

            InventoryCheckResult result = new InventoryCheckResult();
            result.setOk(false);
            result.setOrderId(orderId);
            result.setProductId(productId);
            result.setQuantity(quantity);
            result.setStatus("OUT_OF_STOCK");
            result.setMessage("Product is out of stock for requested quantity");
            result.setRemainingStock(productOpt.map(Product::getStock).orElse(0));
            return result;
        }

        // Decrement stock and persist the fulfillment marker atomically in saga step order.
        Product product = productOpt.get();
        product.setStock(product.getStock() - quantity);
        products.save(product);
        fulfillments.save(new Fulfillment(orderId, "DONE"));

        // Emit success event so the order-service transitions the order to COMPLETED.
        Map<String, Object> okEvent = new HashMap<>();
        okEvent.put("event", "INVENTORY_OK");
        okEvent.put("orderId", orderId);
        okEvent.put("productId", productId);
        okEvent.put("quantity", quantity);
        kafka.send(SagaConstants.TOPIC, okEvent);
        log.info("Fulfilled order {}", orderId);

        InventoryCheckResult result = new InventoryCheckResult();
        result.setOk(true);
        result.setOrderId(orderId);
        result.setProductId(productId);
        result.setQuantity(quantity);
        result.setStatus("FULFILLED");
        result.setMessage("Inventory reserved successfully");
        result.setRemainingStock(product.getStock());
        return result;
    }

    /**
     * Route an incoming saga event to the appropriate handler.
     *
     * <p>Logic: ORDER_CREATED triggers direct fulfillment (simplified choreography without
     * a separate payment step); PAYMENT_CAPTURED triggers fulfillment after payment succeeds.
     *
     * @param event deserialized saga event map from Kafka
     */
    public void handleSagaEvent(Map<String, Object> event) {
        String type = (String) event.get("event");
        if (type == null) {
            return;
        }
        // Both ORDER_CREATED and PAYMENT_CAPTURED trigger an inventory reservation attempt.
        if ("ORDER_CREATED".equals(type) || "PAYMENT_CAPTURED".equals(type)) {
            Long orderId = ((Number) event.get("orderId")).longValue();
            Long productId = ((Number) event.get("productId")).longValue();
            Integer quantity = ((Number) event.get("quantity")).intValue();
            tryFulfill(orderId, productId, quantity);
        }
    }
}

File: 5-saga-orchestration-and-choreography/2-csharp/inventory-service/StockService.cs
// Core business logic — tryFulfill mirrors TS StockService.tryFulfill.
// Reads/writes products + fulfillments from MongoDB; emits INVENTORY_OK / INVENTORY_OUT_OF_STOCK.

using Confluent.Kafka;
using MongoDB.Driver;
using System.Text.Json;

namespace InventoryService;

/// <summary>
/// Encapsulates saga inventory step logic.
/// Checks stock in MongoDB, decrements atomically, records fulfillment for idempotency,
/// and emits the result event back to the Kafka saga topic.
/// </summary>
public sealed class StockService(
    IMongoCollection<Product> products,
    IMongoCollection<Fulfillment> fulfillments,
    IProducer<string, string> producer,
    ILogger<StockService> logger)
{
    private const string Topic = "saga.demo.events";

    /// <summary>
    /// Attempt to reserve <paramref name="quantity"/> units of product <paramref name="productId"/> for order <paramref name="orderId"/>.
    /// Emits INVENTORY_OK on success or INVENTORY_OUT_OF_STOCK on failure.
    /// Idempotent: duplicate calls for the same orderId return ALREADY_FULFILLED without re-emitting.
    /// </summary>
    /// <param name="orderId">Saga order identifier.</param>
    /// <param name="productId">Product to decrement stock for.</param>
    /// <param name="quantity">Units requested.</param>
    /// <returns>Result describing fulfillment outcome.</returns>
    public async Task<InventoryCheckResult> TryFulfillAsync(int orderId, int productId, int quantity)
    {
        // Idempotency check: if already fulfilled, skip duplicate processing.
        var existing = await fulfillments.Find(f => f.OrderId == orderId).FirstOrDefaultAsync();
        if (existing is not null)
        {
            return new InventoryCheckResult(true, orderId, productId, quantity,
                "ALREADY_FULFILLED", "Order has already been fulfilled");
        }

        // Validate stock before mutating inventory state.
        var product = await products.Find(p => p.Id == productId).FirstOrDefaultAsync();
        if (product is null || product.Stock < quantity)
        {
            // Publish compensation trigger so upstream services can rollback/cancel.
            await EmitAsync(new
            {
                @event    = "INVENTORY_OUT_OF_STOCK",
                orderId,
                productId,
            });
            logger.LogInformation("Out of stock for order {OrderId}", orderId);

            return new InventoryCheckResult(false, orderId, productId, quantity,
                "OUT_OF_STOCK", "Product is out of stock for requested quantity",
                product?.Stock ?? 0);
        }

        // Commit stock decrement and fulfillment marker in saga step order.
        var newStock = product.Stock - quantity;
        await products.UpdateOneAsync(
            p => p.Id == productId,
            Builders<Product>.Update.Set(p => p.Stock, newStock));

        await fulfillments.InsertOneAsync(new Fulfillment { OrderId = orderId, Status = "DONE" });

        // Publish success so order-service can mark order COMPLETED.
        await EmitAsync(new
        {
            @event    = "INVENTORY_OK",
            orderId,
            productId,
            quantity,
        });
        logger.LogInformation("Fulfilled order {OrderId}", orderId);

        return new InventoryCheckResult(true, orderId, productId, quantity,
            "FULFILLED", "Inventory reserved successfully", newStock);
    }

    /// <summary>
    /// Handle an incoming saga event — both ORDER_CREATED (simplified two-service
    /// choreography, no separate payment step) and PAYMENT_CAPTURED (full saga flow)
    /// trigger an inventory reservation attempt. Mirrors the Java inventory track.
    /// </summary>
    /// <param name="evtType">Event discriminator string.</param>
    /// <param name="orderId">Order identifier carried in the event.</param>
    /// <param name="productId">Product identifier carried in the event.</param>
    /// <param name="quantity">Quantity carried in the event.</param>
    public async Task HandleSagaEventAsync(string evtType, int orderId, int productId, int quantity)
    {
        // ORDER_CREATED (this C# track has no payment-service) or PAYMENT_CAPTURED
        // both trigger the reserve-stock step in the choreography flow.
        if (evtType == "ORDER_CREATED" || evtType == "PAYMENT_CAPTURED")
        {
            await TryFulfillAsync(orderId, productId, quantity);
        }
    }

    // Serialize and produce a saga event message.
    private async Task EmitAsync(object payload)
    {
        var json = JsonSerializer.Serialize(payload);
        await producer.ProduceAsync(Topic, new Message<string, string> { Value = json });
    }
}
